query/src/sql/planner/binder/select.rs

```diff
@@ -55,6 +60,10 @@ impl Binder {
         // Output of current `SELECT` statement.
         let mut output_context = self.normalize_select_list(&stmt.select_list, &input_context)?;
 
+        if !stmt.group_by.is_empty() {
```

Comment:
The `output_context` is `BindContext` generated by current query, which is actually the state after executing `GROUP BY`.

This is a tricky part in SQL. The aggregate functions look to be defined in `SELECT` clause, but they are actually implicitly defined with `GROUP BY` and get referenced in `SELECT` clause. 

To be more specific, an aggregage query:
```sql
select sum(a) from t group by a;
```
Can be represented in a more understandable way:
```sql
SELECT sum(a) FROM t GROUP BY a AGGREGATE INTO sum(a); 
```

I think what you would like to do is:
1. Extract aggregate functions from `output_context`
2. Resolve `GROUP BY` clause with `input_context` and the aggregate information

My suggestion is that you can add another method like `analyze_aggregate(input_context, output_context)`, which will extract aggregate functions from `output_context` and add them to input_context. Then you can `bind_group_by` with the `input_context` with aggregate information.

Reply:
> I think what you would like to do is:

Yes, you are right!

> My suggestion is that you can add another method like `analyze_aggregate(input_context, output_context)`, which will extract aggregate functions from `output_context` and add them to input_context. Then you can `bind_group_by` with the `input_context` with aggregate information.

Nice suggestion~

> This is a tricky part in SQL. The aggregate functions look to be defined in `SELECT` clause, but they are actually implicitly defined with `GROUP BY` and get referenced in `SELECT` clause.

I printed out ast from parser and tried to see where the aggregation function appears, but I didn't find aggregation functions in group by. At which stage does this implicit conversion occur?

Reply:
> I printed out ast from parser and tried to see where the aggregation function appears, but I didn't find aggregation functions in group by. At which stage does this implicit conversion occur?

Well, it's exactly what you are working on. 🤣